source/slang/ir.cpp

```diff
@@ -3784,9 +3790,183 @@ namespace Slang
         writer->flush();
     }
 
-    //
-    //
-    //
+    // Pre-declare
+    static bool _isTypeOperandEqual(IRInst* a, IRInst* b);
+
+    static bool _areTypeOperandsEqual(IRInst* a, IRInst* b)
+    {
+        // Must have same number of operands
+        const auto operandCountA = Index(a->getOperandCount());
+        if (operandCountA != Index(b->getOperandCount()))
+        {
+            return false;
+        }
+
+        // All the operands must be equal
+        for (Index i = 0; i < operandCountA; ++i)
+        {
+            IRInst* operandA = a->getOperand(i);
+            IRInst* operandB = b->getOperand(i);
+
+            if (!_isTypeOperandEqual(operandA, operandB))
+            {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    static bool _hasNominalEquality(IROp op)
+    {
+        // True if the type should be handled 'nominally' for equality
+        switch (op)
+        {
+            case kIROp_StructType:
+            case kIROp_InterfaceType:
+            case kIROp_Generic:
+            case kIROp_Param:
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    static bool _isNominallyEqual(IRInst* a, IRInst* b)
+    {
+        // Two instruction are nominally equal if their instruction pointer is equal
+        return a == b;
+    }
+
+    // True if a type operand is equal. Operands are 'IRInst' - but it's only a restricted set
+    // if equality of nominal types is by names alone. 
+    static bool _isTypeOperandEqual(IRInst* a, IRInst* b)
+    {
+        if (a == b)
+        {
+            return true;
+        }
+
+        if (a == nullptr || b == nullptr)
+        {
+            return false;
+        }
+
+        IROp opA = IROp(a->op & kIROpMeta_PseudoOpMask);
+        IROp opB = IROp(b->op & kIROpMeta_PseudoOpMask);
+
+        if (opA != opB)
+        {
+            return false;
+        }
+
+        // If it's a constant...
+        if (IRConstant::isaImpl(opA))
+        {
+            // TODO: This is contrived in that we want two types that are the same, but are different
+            // pointers to match here.
+            // If we make GetHashCode for IRType* compatible with isTypeEqual, then we should probably use that.
+            return static_cast<IRConstant*>(a)->isValueEqual(*static_cast<IRConstant*>(b)) &&
+                isTypeEqual(a->getFullType(), b->getFullType());
+        }
+
+        // If it's a type
+        if (IRType::isaImpl(opA))
+        {
+            return isTypeEqual(static_cast<IRType*>(a), static_cast<IRType*>(b));
+        }
+
+        if (_hasNominalEquality(opA))
+        {
+            return _isNominallyEqual(a, b);
+        }
+
+        SLANG_ASSERT(!"Unhandled comparison");
+
+        // We can't equate any other type..
+        return false;
+    }
+
+
+    bool isTypeEqual(IRType* a, IRType* b)
```

Comment:
I might be missing something, but why not just have `isTypeEqual` delgate to `_isTypeOperandEqual` since they seem to perform the exact same tests, with some very minor differences that don't seem necessary. Why not have one function that is the best equality test for instructions that we can manage for now, instead of two?

Reply:
They are slightly different. I already know they are IRType, whereas with the operands they are IRInst and may or may not be IRType. So I (hopefully) have a complete implementation of IRType* but I handle the restricted case of IRInst* that I can handle to make IRType* testing work.

Reply:
What I'm getting at is that I think it would Just Work if you had:

```
bool isTypeEqual(IRType* a, IRType* b)
{
    return _isTypeOperandEqual(a, b);
}
```

For correctness, the `_isTypeOperandEqual` function already has to handle any type you might throw at it (e.g., to know if `Array<X>` and `Array<Y>` are equal it needs to trust `_isTypeOperandEqual(X,Y)` to do the right thing), so it should be a strict generalization of whatever you might write for `isTypeEqual` in the first place.

It seems to me that the shorter definition of `isTypeEqual` would be guaranteed correct, and easier to maintain.